### ISSUE TITLE
Optimization -O2 makes JavaScript throw exception abort()

### ISSUE THREAD

**Issue by bd339** [emscripten-core/emscripten#3858]
I have compiled this SDL library:
https://github.com/AMDmi3/libSDL2pp
And tested it with this code:
https://gist.github.com/bd339/10c69b69d66e6bf07e91
When the code is built as follows the file size is correctly displayed and no exceptions are thrown:
`emcc -I/home/benjamin/libSDL2pp -I/home/benjamin/.emscripten_cache/ports-builds/sdl2/include -L/home/benjamin/libSDL2pp -lSDL2pp -std=c++11 -Wall -o readfile.html --preload-file /path/to/file readfile.cpp`
If the library or the test application or both are built with -O2 the application will throw an exception when run in the browser (after the file size is displayed).
If I build just the library with -O2 as follows the code run in the browser will not be minified and thus the stack trace makes more sense:
`emcmake cmake -D SDL2_INCLUDE_DIR:PATH=/home/benjamin/.emscripten_cache/ports-builds/sdl2/include -D SDL2_LIBRARY:PATH=/home/benjamin/.emscripten_cache/ports-builds/sdl2/libsdl2.bc -D CMAKE_CXX_FLAGS:STRING=-O2 .`
`emmake make`
When the resulting binary is linked with readfile.cpp as shown in the first build command, this is the output I see in the JavaScript console:
https://gist.github.com/bd339/24f72c0ba1ad1d7d1b45
To get the library to build (with and without optimizations) I had to disable SDL_image & SDL_ttf & SDL_mixer support as well as disable examples & tests & live tests.
The code was run in Firefox in Linux Mint 64 bit and emscripten was compiled from source.

**Comment by kripken** [comment]
Might help to try [these compiler settings](http://kripken.github.io/emscripten-site/docs/porting/Debugging.html#compiler-settings), they can find some errors.

**Comment by bd339** [comment]
I have tried `ASSERTIONS=1` (and 2), `SAFE_HEAP=1`, `SAFE_HEAP_LOG=1` and `DEMANGLE_SUPPORT=1` but none of them have yielded any particularly useful information. Everything seems to run fine until it suddenly throws revealing just the output I posted (https://gist.github.com/bd339/24f72c0ba1ad1d7d1b45)

**Comment by kripken** [comment]
It's possible you are hitting an SDL bug, given that stack trace. Might be interesting to log out all mallocs and frees and see if it's doing a double-free or something like that.

**Comment by bd339** [comment]
I tried to add logs to free and malloc and the results indicate that a lot of memory allocated in the unoptimized builds are never freed again, despite the program terminating without errors. I also noticed one memory address being freed (smaller than the rest of the memory addresses) in the optimized build before it is ever reported to be allocated by malloc.
These were the results (unoptimized):

succesfully malloced 5271560 readmdlfile-noop.js:32091:4
succesfully malloced 5272336 readmdlfile-noop.js:30849:3
succesfully malloced 5272352 readmdlfile-noop.js:30849:3
succesfully malloced 5272368 readmdlfile-noop.js:30849:3
succesfully malloced 5272392 readmdlfile-noop.js:30849:3
succesfully malloced 5272408 readmdlfile-noop.js:30849:3
succesfully malloced 5272448 readmdlfile-noop.js:30849:3
about to free not-null 5272408

And these were the results from the optimized build:

succesfully malloced 5271560 readmdlfile.js:32094:4
succesfully malloced 5272336 readmdlfile.js:30852:3
succesfully malloced 5272352 readmdlfile.js:30852:3
succesfully malloced 5272368 readmdlfile.js:30852:3
succesfully malloced 5272392 readmdlfile.js:30852:3
succesfully malloced 5272408 readmdlfile.js:30852:3
succesfully malloced 5272448 readmdlfile.js:30852:3
about to free not-null 5272408 readmdlfile.js:51667:2
about to free not-null 19696 readmdlfile.js:51667:2
succesfully malloced 5272488 readmdlfile.js:30852:3
succesfully malloced 5272408 readmdlfile.js:29880:6
succesfully malloced 5272544 readmdlfile.js:30852:3
succesfully malloced 5272424 readmdlfile.js:30832:3
succesfully malloced 5272584 readmdlfile.js:30852:3
succesfully malloced 5272624 readmdlfile.js:30852:3
succesfully malloced 5272664 readmdlfile.js:30852:3
about to free not-null 5272544 readmdlfile.js:51667:2
about to free not-null 5272624 readmdlfile.js:51667:2
succesfully malloced 5272736 readmdlfile.js:30852:3
about to free not-null 5272584 readmdlfile.js:51667:2
about to free not-null 5272736 readmdlfile.js:51667:2
succesfully malloced 5272544 readmdlfile.js:29880:6
about to free not-null 5272424 readmdlfile.js:51667:2
about to free not-null 5272664 readmdlfile.js:51667:2
about to free not-null 5272488 readmdlfile.js:51667:2
about to free not-null 5272408 readmdlfile.js:51667:2
about to free not-null 5272544 readmdlfile.js:51667:2
succesfully malloced 5272408 readmdlfile.js:29756:5
succesfully malloced 5272488 readmdlfile.js:30852:3
succesfully malloced 5272504 readmdlfile.js:30852:3
succesfully malloced 5272544 readmdlfile.js:30852:3
succesfully malloced 5272568 readmdlfile.js:30852:3
about to free not-null 5272568 readmdlfile.js:51667:2
succesfully malloced 5272568 readmdlfile.js:30852:3
about to free not-null 5272544 readmdlfile.js:51667:2
about to free not-null 5272504 readmdlfile.js:51667:2
about to free not-null 5272408 readmdlfile.js:51667:2
about to free not-null 5272488 readmdlfile.js:51667:2
about to free not-null 5272568

After the last free the usual stack trace appears. Notice especially the (odd) free `about to free not-null 19696 readmdlfile.js:51667:2` towards the beginning.
